# 1. Data wrangling, cleaning and summary
** **
## Table of contents:

1. Importing libraries <br>
2. Importing dataframe <br>
3. Data wrangling
    - 3.1 Dropping columns
    - 3.2 Checking datatypes
    - 3.3 Round decimal numbers
    - 3.4 Checking value counts and missing values
4. Consistency checks
    - 4.1 Check for mixed-type data columns
    - 4.2 Check for duplicates
    - 4.3 Summary of missing values
5. Descriptive statistics
6. Deriving a new variable
7. Exporting dataframes

# 1. Importing libraries
** **

In [1]:
import pandas as pd
import numpy as np
import os

# 2. Importing dataframe
** **

In [2]:
# Creating a path variabile for the folder
path = r'C:\Users\Simone\Desktop\Career Foundry\Esercizi modulo 7\CO2 emissions analysis'

In [3]:
# Importing the dataframe without index column
df_co2 = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'emissions.csv'), index_col = False)

# 3. Data wrangling
** **

In [4]:
# Checking the shape
df_co2.shape

(55440, 11)

In [5]:
# Printing the first 5 rows
df_co2.head()

Unnamed: 0 Country                Energy_type  Year  Energy_consumption  \
0           0   World           all_energy_types  1980          292.899790   
1           1   World                       coal  1980           78.656134   
2           2   World                natural_gas  1980           53.865223   
3           3   World  petroleum_n_other_liquids  1980          132.064019   
4           4   World                    nuclear  1980            7.575700   

   Energy_production           GDP    Population  Energy_intensity_per_capita  \
0         296.337228  27770.910281  4.298127e+06                    68.145921   
1          80.114194  27770.910281  4.298127e+06                    68.145921   
2          54.761046  27770.910281  4.298127e+06                    68.145921   
3         133.111109  27770.910281  4.298127e+06                    68.145921   
4           7.575700  27770.910281  4.298127e+06                    68.145921   

   Energy_intensity_by_GDP  CO2_emission  
0                   10.547   4946.627130  
1                   10.547   1409.790188  
2                   10.547   1081.593377  
3                   10.547   2455.243565  
4                   10.547      0.000000

We don't need the first column "Unnamed", we already have an index.

## 3.1 Dropping columns

In [6]:
# Removing the column not necessary
df_co2 = df_co2.drop(columns = ['Unnamed: 0'])

In [7]:
# Checking the result
df_co2.head()

Country                Energy_type  Year  Energy_consumption  \
0   World           all_energy_types  1980          292.899790   
1   World                       coal  1980           78.656134   
2   World                natural_gas  1980           53.865223   
3   World  petroleum_n_other_liquids  1980          132.064019   
4   World                    nuclear  1980            7.575700   

   Energy_production           GDP    Population  Energy_intensity_per_capita  \
0         296.337228  27770.910281  4.298127e+06                    68.145921   
1          80.114194  27770.910281  4.298127e+06                    68.145921   
2          54.761046  27770.910281  4.298127e+06                    68.145921   
3         133.111109  27770.910281  4.298127e+06                    68.145921   
4           7.575700  27770.910281  4.298127e+06                    68.145921   

   Energy_intensity_by_GDP  CO2_emission  
0                   10.547   4946.627130  
1                   10.547   1409.790188  
2                   10.547   1081.593377  
3                   10.547   2455.243565  
4                   10.547      0.000000

In [8]:
# Checking also the tail
df_co2.tail()

Country                Energy_type  Year  Energy_consumption  \
55435  Zimbabwe                       coal  2019            0.045064   
55436  Zimbabwe                natural_gas  2019            0.000000   
55437  Zimbabwe  petroleum_n_other_liquids  2019            0.055498   
55438  Zimbabwe                    nuclear  2019                 NaN   
55439  Zimbabwe         renewables_n_other  2019            0.068089   

       Energy_production      GDP  Population  Energy_intensity_per_capita  \
55435           0.075963  37.6204     14654.2                    11.508701   
55436           0.000000  37.6204     14654.2                    11.508701   
55437           0.000000  37.6204     14654.2                    11.508701   
55438                NaN  37.6204     14654.2                    11.508701   
55439           0.067499  37.6204     14654.2                    11.508701   

       Energy_intensity_by_GDP  CO2_emission  
55435                 4.482962      4.586869  
55436                 4.482962      0.000000  
55437                 4.482962      4.377890  
55438                 4.482962      0.000000  
55439                 4.482962      0.000000

## 3.2 Checking datatypes

In [9]:
# Checking columns datatypes
df_co2.dtypes

Country                         object
Energy_type                     object
Year                             int64
Energy_consumption             float64
Energy_production              float64
GDP                            float64
Population                     float64
Energy_intensity_per_capita    float64
Energy_intensity_by_GDP        float64
CO2_emission                   float64
dtype: object

Datatypes are correct.

## 3.3 Round decimal numbers

A lot of floating columns have too longs decimal numbers. I will round them to 2 decimal places. <br>
Population will be instead rounded to 0 decimal places.

In [10]:
df_co2.Energy_consumption = df_co2.Energy_consumption.round(2)

In [11]:
df_co2.Energy_production = df_co2.Energy_production.round(2)

In [12]:
df_co2.GDP = df_co2.GDP.round(2)

In [13]:
df_co2.Population = df_co2.Population.round()

In [14]:
df_co2.Energy_intensity_per_capita = df_co2.Energy_intensity_per_capita.round(2)

In [15]:
df_co2.Energy_intensity_by_GDP = df_co2.Energy_intensity_by_GDP.round(2)

In [16]:
df_co2.CO2_emission = df_co2.CO2_emission.round(2)

In [17]:
# Checking the result
df_co2.head(10)

Country                Energy_type  Year  Energy_consumption  \
0        World           all_energy_types  1980              292.90   
1        World                       coal  1980               78.66   
2        World                natural_gas  1980               53.87   
3        World  petroleum_n_other_liquids  1980              132.06   
4        World                    nuclear  1980                7.58   
5        World         renewables_n_other  1980               20.70   
6  Afghanistan           all_energy_types  1980                0.03   
7  Afghanistan                       coal  1980                0.00   
8  Afghanistan                natural_gas  1980                0.00   
9  Afghanistan  petroleum_n_other_liquids  1980                0.01   

   Energy_production       GDP  Population  Energy_intensity_per_capita  \
0             296.34  27770.91   4298127.0                        68.15   
1              80.11  27770.91   4298127.0                        68.15   
2              54.76  27770.91   4298127.0                        68.15   
3             133.11  27770.91   4298127.0                        68.15   
4               7.58  27770.91   4298127.0                        68.15   
5              20.78  27770.91   4298127.0                        68.15   
6               0.07       NaN     13356.0                         1.99   
7               0.00       NaN     13356.0                         1.99   
8               0.06       NaN     13356.0                         1.99   
9               0.00       NaN     13356.0                         1.99   

   Energy_intensity_by_GDP  CO2_emission  
0                    10.55       4946.63  
1                    10.55       1409.79  
2                    10.55       1081.59  
3                    10.55       2455.24  
4                    10.55          0.00  
5                    10.55          0.00  
6                     0.00           NaN  
7                     0.00           NaN  
8                     0.00           NaN  
9                     0.00           NaN

All variables have been rounded correctly.

## 3.4 Checking value counts and missing values

In [18]:
# Removing the limit to max rows displayed
pd.set_option('display.max_rows', None)

In [19]:
df_co2['Country'].value_counts(dropna = False)

World                           240
Netherlands Antilles            240
New Zealand                     240
Nicaragua                       240
Niger                           240
Nigeria                         240
Niue                            240
North Korea                     240
North Macedonia                 240
Northern Mariana Islands        240
Norway                          240
Oman                            240
Pakistan                        240
Palestinian Territories         240
Panama                          240
Papua New Guinea                240
Paraguay                        240
Peru                            240
Philippines                     240
Poland                          240
Portugal                        240
Puerto Rico                     240
Qatar                           240
Reunion                         240
Romania                         240
Russia                          240
Rwanda                          240
New Caledonia               

The variable contains 231 unique values, each of them recorded in 240 observations.

In [20]:
df_co2['Energy_type'].value_counts(dropna = False)

all_energy_types             9240
coal                         9240
natural_gas                  9240
petroleum_n_other_liquids    9240
nuclear                      9240
renewables_n_other           9240
Name: Energy_type, dtype: int64

The variable contains 6 unique values, each of them recorded in 9240 observations.

In [21]:
df_co2['Year'].value_counts(dropna = False)

1980    1386
1981    1386
2002    1386
2003    1386
2004    1386
2005    1386
2006    1386
2007    1386
2008    1386
2009    1386
2010    1386
2011    1386
2012    1386
2013    1386
2014    1386
2015    1386
2016    1386
2017    1386
2018    1386
2001    1386
2000    1386
1999    1386
1989    1386
1982    1386
1983    1386
1984    1386
1985    1386
1986    1386
1987    1386
1988    1386
1990    1386
1998    1386
1991    1386
1992    1386
1993    1386
1994    1386
1995    1386
1996    1386
1997    1386
2019    1386
Name: Year, dtype: int64

The variable contains 40 unique values (1980-2019), each of them recorded in 1386 observations.

In [22]:
df_co2['Energy_consumption'].value_counts(dropna = False)

 0.00      17722
 NaN       11153
 0.01       3586
 0.02       1900
 0.03       1486
 0.04       1215
 0.05        977
 0.06        738
 0.07        593
 0.08        553
 0.09        483
 0.10        460
 0.11        456
 0.12        417
 0.13        399
 0.14        339
 0.15        331
 0.17        300
 0.19        293
 0.16        276
 0.18        265
 0.20        231
 0.22        210
 0.23        191
 0.21        186
 0.24        142
 0.32        140
 0.28        140
 0.30        140
 0.34        136
 0.26        132
 0.25        131
 0.29        129
 0.27        121
 0.35        121
 0.33        120
 0.31        117
 0.37        116
 0.36        114
 0.41        112
 0.38        108
 0.40        104
 0.39        102
 0.45         95
 0.53         88
 0.47         88
 0.48         86
 0.44         85
 0.42         85
 0.43         83
 0.46         75
 0.54         72
 0.52         72
 0.69         72
 0.51         71
 0.50         71
 0.55         69
 0.59         69
 0.56         

The variable contains 11153 missing values (NaN) and 17722 "0".

In [23]:
df_co2['Energy_production'].value_counts(dropna = False)

0.00      25497
NaN       11151
0.01       2210
0.02       1117
0.03        841
0.04        769
0.05        605
0.06        494
0.07        427
0.08        391
0.10        335
0.09        329
0.12        278
0.13        271
0.11        267
0.14        250
0.15        227
0.17        203
0.16        201
0.18        167
0.19        162
0.20        150
0.22        129
0.21        121
0.24        119
0.23        116
0.26        101
0.37         99
0.25         99
0.29         98
0.28         94
0.27         88
0.33         86
0.38         84
0.30         83
0.32         76
0.46         76
0.36         75
0.40         72
0.31         72
0.50         68
0.34         68
0.43         68
0.47         67
0.39         67
0.51         66
0.42         66
0.35         65
0.53         65
0.41         63
0.45         62
0.44         59
0.52         58
0.55         58
0.54         57
0.57         56
0.56         55
0.61         55
0.48         50
0.59         50
0.49         47
0.71         47
0.66    

The variable contains 11151 missing values (NaN) and 25497 "0".

In [24]:
df_co2['GDP'].value_counts(dropna = False)

NaN          15414
0.80            78
0.64            72
0.43            60
1.25            60
0.49            54
0.50            54
0.67            48
0.79            48
1.24            48
2.92            42
0.51            42
0.14            42
0.63            42
0.17            42
1.26            42
0.57            42
1.27            42
0.81            42
0.39            42
1.09            36
0.91            36
1.10            36
0.94            36
2.02            36
1.22            36
1.02            36
4.75            36
1.93            36
177.63          36
3.05            36
1.34            36
2.46            36
1.44            30
0.55            30
0.72            30
2.42            30
4.53            30
0.95            30
2.35            30
2.81            30
1.76            30
4.39            30
1.23            30
0.40            30
0.53            30
0.52            30
1.05            30
13.53           30
0.78            30
2.06            30
0.37            30
1.91        

The variable contains 15414 missing values (NaN).

In [25]:
df_co2['Population'].value_counts(dropna = False)

NaN          9426
56.0          180
71.0          144
57.0          138
108.0         126
101.0          96
70.0           90
103.0          90
102.0          84
104.0          78
96.0           78
105.0          78
109.0          78
97.0           72
63.0           66
106.0          60
55.0           60
94.0           60
66.0           60
107.0          54
65.0           54
64.0           48
111.0          48
98.0           48
54.0           48
366.0          48
62.0           48
95.0           48
100.0          42
158.0          42
159.0          42
58.0           42
46.0           42
400.0          42
20.0           42
29.0           42
160.0          42
41.0           36
93.0           36
162.0          36
376.0          36
150.0          36
404.0          36
51.0           36
91.0           36
110.0          36
72.0           36
337.0          36
398.0          36
99.0           36
52.0           36
75.0           36
378.0          30
284.0          30
386.0          30
19.0      

The variable contains 9426 missing values (NaN).

In [26]:
df_co2['Energy_intensity_per_capita'].value_counts(dropna = False)

0.00       5784
NaN        5082
1.52         60
1.00         54
1.94         54
1.10         54
1.44         54
2.10         54
2.76         54
0.99         54
1.48         54
0.98         54
1.37         48
1.14         48
1.76         48
2.07         48
0.97         48
1.54         48
1.56         48
1.58         48
1.47         42
3.43         42
3.22         42
2.69         42
1.12         42
1.98         42
1.60         42
2.12         42
1.50         42
2.81         42
1.57         42
1.09         36
3.25         36
7.06         36
1.49         36
3.31         36
1.80         36
2.56         36
3.38         36
3.20         36
1.99         36
0.79         36
0.87         36
2.27         36
2.39         36
3.41         36
2.57         36
0.94         36
2.52         36
7.67         36
1.25         36
7.85         36
3.46         36
0.88         36
0.74         30
1.05         30
1.06         30
14.13        30
1.65         30
2.04         30
6.69         30
3.47         30
3.29    

The variable contains 5082 missing values (NaN) and 5784 "0" values.

In [27]:
df_co2['Energy_intensity_by_GDP'].value_counts(dropna = False)

0.00      11442
NaN        5082
3.38        126
3.37        120
4.05        114
1.26        114
1.33        114
3.33        114
4.36        108
1.58        108
3.66        108
2.80        102
4.29        102
2.97        102
3.02        102
3.03        102
1.41        102
2.94        102
3.28        102
3.57        102
2.95         96
3.00         96
2.62         96
3.54         96
1.72         96
2.74         96
1.28         96
2.56         96
1.76         96
2.93         96
3.59         96
3.06         96
3.46         96
1.87         96
3.20         96
3.48         96
1.20         96
4.33         96
1.59         90
3.31         90
3.08         90
1.68         90
2.01         90
2.99         90
1.75         90
2.03         90
3.97         90
2.70         90
3.43         90
2.40         90
1.84         90
2.54         90
3.49         90
4.25         90
2.96         90
3.30         90
4.08         90
4.18         90
1.81         90
1.39         84
4.83         84
2.88         84
3.23    

The variable contains 5082 missing values (NaN) and 11442 "0" values.

In [28]:
df_co2['CO2_emission'].value_counts(dropna = False)

 0.00        27359
 NaN          3826
 0.20          541
 0.10          451
 0.30          369
 0.01          269
 0.70          251
 0.50          247
 0.40          238
 0.60          219
 0.03          204
 0.04          198
 1.00          175
 0.08          174
 0.02          170
 0.06          159
 0.80          153
 1.10          147
 1.30          144
 0.05          140
 1.70          138
 0.07          132
 0.90          128
 1.20          128
 0.09          103
 2.40           95
 2.00           90
 1.90           89
 1.40           87
 1.50           81
 2.10           80
 11.00          80
 12.00          79
 1.80           78
 3.20           78
 1.60           72
 2.30           72
 2.20           67
 2.50           67
 3.00           66
 2.70           66
 0.15           66
 2.60           61
 2.90           60
 0.17           58
 0.14           58
 0.18           57
 0.16           56
 0.11           56
 0.19           56
 0.12           55
 0.27           54
 14.00      

The variable contains 3826 missing values (NaN) and 27359 "0" values.

<b> Observations: </b> <br>
We found a lot of missing values. Additionally, there are a lot of 0 values too. However, this doesn't mean they're actually missing values. Probably numbers just didn't reach the minimum threshold (CO2 emissions are measured in millions of tonnes).

# 4. Consistency checks
** **

## 4.1 Check for mixed-type data columns

In [29]:
# Checking if there are mixed type columns
for col in df_co2.columns.tolist():
  weird = (df_co2[[col]].applymap(type) != df_co2[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_co2[weird]) > 0:
    print (col)

0 columns printed. 0 mixed-type columns detected.

## 4.2 Check for duplicates

In [30]:
# Checking duplicates
df_dups = df_co2[df_co2.duplicated()]

In [31]:
# Testing
df_dups.head(10)

Empty DataFrame
Columns: [Country, Energy_type, Year, Energy_consumption, Energy_production, GDP, Population, Energy_intensity_per_capita, Energy_intensity_by_GDP, CO2_emission]
Index: []

0 duplicates found.

## 4.3 Summary of missing values

In [32]:
df_co2.isnull().sum()

Country                            0
Energy_type                        0
Year                               0
Energy_consumption             11153
Energy_production              11151
GDP                            15414
Population                      9426
Energy_intensity_per_capita     5082
Energy_intensity_by_GDP         5082
CO2_emission                    3826
dtype: int64

The dataframe contains 55440 rows and 10 columns. <br>
In other words, it contains 554400 values. <br>
61.134 values are missing (11% of total).

# 5. Descriptive statistics
** **

In [33]:
# Checking for abnormalities in descriptive statistics
df_co2.describe()

Year  Energy_consumption  Energy_production            GDP  \
count  55440.0000        44287.000000       44289.000000   40026.000000   
mean    1999.5000            1.537627           1.532586     827.144181   
std       11.5435           15.456603          15.303575    5981.703197   
min     1980.0000           -0.160000           0.000000       0.120000   
25%     1989.7500            0.000000           0.000000       9.740000   
50%     1999.5000            0.020000           0.000000      47.760000   
75%     2009.2500            0.210000           0.110000     263.690000   
max     2019.0000          601.040000         611.510000  127690.250000   

         Population  Energy_intensity_per_capita  Energy_intensity_by_GDP  \
count  4.601400e+04                 50358.000000             50358.000000   
mean   6.263020e+04                    71.898907                 3.695095   
std    4.562088e+05                   113.728744                 4.590756   
min    1.100000e+01                     0.000000                 0.000000   
25%    1.142000e+03                     3.800000                 0.900000   
50%    6.158000e+03                    29.780000                 2.990000   
75%    2.004300e+04                    95.520000                 4.970000   
max    7.714631e+06                  1139.320000               166.910000   

       CO2_emission  
count  51614.000000  
mean      78.800072  
std      902.221446  
min       -0.010000  
25%        0.000000  
50%        0.000000  
75%        4.320000  
max    35584.930000

The minimum value in “Energy_consumption” and “CO2_emission” is negative. <br>
There is a huge gap between the minimum value for GDP and the maximum value, hinting at huge differences between poor countries and rich countries. <br>
The same applies to energy and emissions: the data is very spread and there is a great difference in countries energy consumption and emissions. <br>
75% of total values in “Energy_consumption”, “Energy_production” is less than 1. <br>
This will be sooner investigated through charts and deeper analysis.


# 6. Deriving a new variable
** **

Considering how much the GDP seems it will be important in future analysis, I decided to create a flag to divide GDP into four categories, using the quartiles range.

In [34]:
# Creating first condition
df_co2.loc[df_co2['GDP'] <= 9.74, 'GDP_category'] = 'Low'

In [35]:
# Creating second condition
df_co2.loc[(df_co2['GDP'] > 9.74) & (df_co2['GDP'] <= 47.76), 'GDP_category'] = 'Medium'

In [36]:
# Creating third condition
df_co2.loc[(df_co2['GDP'] > 47.76) & (df_co2['GDP'] <= 263.69), 'GDP_category'] = 'Medium-high'

In [37]:
# Creating last condition
df_co2.loc[df_co2['GDP'] > 263.69, 'GDP_category'] = 'High'

In [38]:
# Checking the results
df_co2.head(10)

Country                Energy_type  Year  Energy_consumption  \
0        World           all_energy_types  1980              292.90   
1        World                       coal  1980               78.66   
2        World                natural_gas  1980               53.87   
3        World  petroleum_n_other_liquids  1980              132.06   
4        World                    nuclear  1980                7.58   
5        World         renewables_n_other  1980               20.70   
6  Afghanistan           all_energy_types  1980                0.03   
7  Afghanistan                       coal  1980                0.00   
8  Afghanistan                natural_gas  1980                0.00   
9  Afghanistan  petroleum_n_other_liquids  1980                0.01   

   Energy_production       GDP  Population  Energy_intensity_per_capita  \
0             296.34  27770.91   4298127.0                        68.15   
1              80.11  27770.91   4298127.0                        68.15   
2              54.76  27770.91   4298127.0                        68.15   
3             133.11  27770.91   4298127.0                        68.15   
4               7.58  27770.91   4298127.0                        68.15   
5              20.78  27770.91   4298127.0                        68.15   
6               0.07       NaN     13356.0                         1.99   
7               0.00       NaN     13356.0                         1.99   
8               0.06       NaN     13356.0                         1.99   
9               0.00       NaN     13356.0                         1.99   

   Energy_intensity_by_GDP  CO2_emission GDP_category  
0                    10.55       4946.63         High  
1                    10.55       1409.79         High  
2                    10.55       1081.59         High  
3                    10.55       2455.24         High  
4                    10.55          0.00         High  
5                    10.55          0.00         High  
6                     0.00           NaN          NaN  
7                     0.00           NaN          NaN  
8                     0.00           NaN          NaN  
9                     0.00           NaN          NaN

In [39]:
# Checking the value count
df_co2['GDP_category'].value_counts(dropna = False)

NaN            15414
Low            10014
Medium-high    10008
High           10002
Medium         10002
Name: GDP_category, dtype: int64

This new variable can now be used in future analyses.

# 7. Exporting dataframes
** **

In [40]:
# Exporting the new dataframe (clean and with derived columns) in pkl format
df_co2.to_pickle(os.path.join(path, '02. Data','Prepared Data', 'emissions_clean.pkl'))

To work with more timely data, I am also creating a subset containing only data from last 5 years (from 2015 to 2019).

In [41]:
# Creating the subset
df_co2_2015 = df_co2.loc[df_co2['Year'].isin([2015, 2016, 2017, 2018, 2019])]

In [42]:
# Checking the shape of the new subset
df_co2_2015.shape

(6930, 11)

The new dataframe is considerably shorter (less than 7000 rows).

In [43]:
# Checking number of missing values in the subsetted dataframe
df_co2_2015.isnull().sum()

Country                           0
Energy_type                       0
Year                              0
Energy_consumption             1161
Energy_production              1159
GDP                             828
Population                      810
Energy_intensity_per_capita     360
Energy_intensity_by_GDP         360
CO2_emission                    171
GDP_category                    828
dtype: int64

There is still a considerable amount of missing values.

In [44]:
# Exporting the subsetted dataframe (clean and with derived columns) in pkl format
df_co2_2015.to_pickle(os.path.join(path, '02. Data','Prepared Data', 'emissions_clean_2015.pkl'))